catt/controllers.py

```diff
@@ -447,6 +461,25 @@ def restore(self, data):
                             title=data["title"], thumb=data["thumb"])
 
 
+class DashCastController(CastController):
+    app_ready = "Application ready"
+
+    def __init__(self, cast, name, app_id, prep=None):
+        self._controller = PyChromecastDashCastController()
+        super(DashCastController, self).__init__(cast, name, app_id, prep=prep)
+
+    def load_url(self, url, **kwargs):
+        self._controller.load_url(url, force=True)
+
+    def _prep_app(self):
+        """Make sure desired chromecast app is running."""
+        # we must force the launch of the DashCast app because it, by design,
+        # becomes unresponsive after a website is loaded
+        self._cast.socket_client.receiver_controller.launch_app(self._cast_listener.app_id, force_launch=True)
+        self._cast.start_app(self._cast_listener.app_id)
+        self._cast_listener.app_ready.wait()
```

Comment:
If you clear the ```app_ready``` flag before force launching the app, it would not be necessary to modify ```CastStatusListener``` or use the class variable. Also the ```self._cast.start_app``` call is not needed anymore.

Reply:
Wouldn't not modifying `CastStatusListener` make things rely on a race condition ?
The modified version of `CastStatusListener` guarantees that the app is only ready when it *should* be.

The DashCast app has an "Application Is Starting" state in which it can not yet receive URLs: https://github.com/stestagg/dashcast/blob/master/receiver.html#L163

This is one of the reasons, actually, that I liked that `catt info` showed the state of the DashCast app.

Reply:
> Wouldn't not modifying CastStatusListener make things rely on a race condition ?

Hmm, I guess thats true, although it's extremely unlikely that this would ever be a real problem. I've tested my own suggestion and could not get it to fail.
I'm keen on the code needed to circumvent the odd behaviour of dashcast, being contained within the overridden ```_prep_app```. And I am honestly not keen on the class variable and the use of it in ```CastStatusListener```.

@skorokithakis I'm not completely sure I'm in the right here, any input?

Reply:
I did not like the change either on `CastStatusListener` but either we leave things as is or we patch `DashCast` to behave as expected. Race conditions are must always be avoided because debugging them is a *real* pain. My vote is to do nothing and push the merge button.

Reply:
I'll have to agree with @marcosdiez here, if eliminating a race condition means that the code will be a bit uglier, I can live with that, since it will be really hard to debug something like "Sometimes catt doesn't work" or "catt doesn't work on my new Chromecast".

I would, however, like to add comments here and on `CastStatusListener` explaining exactly which race condition this guards against, and how it does that. I can definitely see us coming back in a year, not remembering this discussion and saying "Oh we don't actually need this, we can just clear the `app_ready` flag" and fall into the race condition trap.

Reply:
The (very) remote possibility of a race condition would only occur when using dashcast, but fair enough.
I still don't understand why the current logic in ```CastStatusListener``` is needed, when we could just do this in the constructor:

```
if app_id == active_app_id and active_app_id != DASHCAST_APP_ID:
    self.app_ready.set()
```
This would have the same effect as my previous suggestion without introducing the risk of a race condition. Am I missing something?

Reply:
Hmm, it's been a while since I visited this part of the code and can't remember the flow here. Can either of you refresh my memory on why the race condition would happen?

Reply:
I see, thank you. I would still be in favor of preventing the race condition, but your proposed solution seems good to me, as I'm not a fan of modifying `CastStatusListener` either. @marcosdiez, do you see any reason why @theychx's suggestion wouldn't work?

Reply:
The whole problem arises because the DashCast app is odd.

It has two idiosyncrasies:

a) once loaded it, by design becomes unresponsive (because it redirects to the casted webpage which has no idea it's being rendered in Chromecast) (this has been solved previously by killing it and now by force loading the app every time)

b) when you load the chromecast app, it's initial state is `Application is Starting` https://github.com/stestagg/dashcast/blob/master/receiver.html#L163 

Only eventually the status becomes `Application Ready` https://github.com/stestagg/dashcast/blob/master/receiver.html#L143

The changes I did in `CastStatusListener` (which in the end of the day are not that ugly), explicitly set the state to ready only when it's really ready (by waiting for the `Application Ready` status).

And since I live in Brazil, which is the other side of earth, the internet has much more latency than in Europe and the US because we are very far and light is not that fast.

Australians have it worse. As soon as `catt` becomes a worldwide hit they will fill most of the bugs if we leave the race condition!

I am completely against leaving a race condition there. It will make casting a url something that will randomly fail, more often if there is more latency and after we forget all this, we will never ever fix it because, as with every race condition, even reproducing is a challenge. We know the problem, we have the solution ready and we are willing not to use it. Sorry to say that, but this is negligence. 

Thinking about it, the discussion here is the most technical solution versus the least intrusive solution. The former is always my choice.

`DashCast` needs special treatment and that's clearly defined in `CastStatusListener`. We can't hide away from that fact. `CastStatusListener` is the right place to deal with it.

Also, any coder in the future will see that. I can, of course, improve the comments, but even if we leave the code as is, I don't believe anybody would ever delete any of the code from `CastStatusListener` by accident in the future, for cleanup's sake. (different from a flag that arbitrarily and unintuitively set's the ready state to false)

Guys, sorry for not agreeing. This is just my opinion and I do respect if you decide otherwise. Please don't get offended by me not agreeing.

Reply:
Hey man, nobody is offended, we're just discussing code. Let me read the rest of your comment and reply in more detail.

Reply:
@marcosdiez What is the status of other applications when they become ready?

Reply:
So here are my thoughts:

1. I'm really averse to having the race condition, I think we're all in the same page that we want to avoid that, even if the code gets a bit uglier.
1. If the other applications have the same status string as DashCast, maybe we can generalize and not hardcode it for just DashCast.
2. Since the `app_ready` string is a DashCast-specific hack, I'm not crazy about it being called `app_ready` (collides with another variable and the person reading the code gets confused). I would rather just have the code in `_is_app_ready` be `return status.status_text == "Application ready"` with a big comment saying what you said above about the race condition.
3. I'm still not sure why @theychx's suggestion wouldn't work. Can you elaborate a bit, @marcosdiez?

Other than that, I think this is very close to merging, great work!

Reply:
@marcosdiez I think you missed my new proposal, which do not introduce the risk of a race condition. Please read that, and explain to us why that would not suffice. I have not been able to get this new suggestion to fail either.
Also, I just tested the signals coming from pychromecast to the listener, and it looks as though the app id updates don't fire before the app is truly ready. But I urge you to try and make my second suggestion fail, and if you succeed I will of course change my mind.

PS: I'm not offended. :smiley:

Reply:
@theychx To clarify, you mean we'd add that check to the `__init__` of `DashCastController`, right?

Reply:
No, __init__ of ```CastStatusListener```.
@marcosdiez is right about the app id update firing while the status text is still "Application is starting". I'm just not a fan of remedying that, unless it becomes a real problem for users. Like I said, my second suggestion seems perfectly reliable.

Reply:
@skorokithakis 
> If the other applications have the same status string as DashCast, maybe we can generalize and not hardcode it for just DashCast.

No other app I have tested has this behaviour.

Reply:
Also, it looks like your suggestion sets the `app_ready` flag if the app is not `DashCast`, but don't we need a way to tell that DashCast is ready? Sorry to be such a pain, I just don't understand the flow here exactly.

Reply:
Yep, but he also implemented the extra status text check, and it is the necessity of this I am contesting.

Reply:
@theychx Can we discuss somewhere more real-time? I've created a Gitter room and invited you there (you can just log in with Github).

Reply:
@marcosdiez You're welcome too, of course.

Reply:
Hey @marcosdiez, @theychx and I discussed this on Gitter, and we agree that eliminating the race condition makes the most sense. However, could you change the code in the check to this?:

```python
def _is_app_ready(self, status):
    if status.app_id != self.app_id:
        return False
    if self.app_id == DASHCAST_APP_ID:
        # Big-ass comment about why this is here here.
        return status.status_text == "Application ready"
    return True
```

That way, we get rid of the confusingly-named class variable that's only used in one place, and we have a comment on the special case explaining exactly why it happens.

Also, could you add a comment on the check in `__init__` on the `CastStatusListener` about why the DashCast check is necessary? I think this is a good compromise between clearly documenting the special-case code and eliminating any potential race conditions.